flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/FileChannelBoundedData.java

```diff
@@ -0,0 +1,160 @@
+ */
+final class FileChannelBoundedData implements BoundedData {
+
+	private final Path filePath;
+
+	private final FileChannel fileChannel;
+
+	private final ByteBuffer[] headerAndBufferArray;
+
+	private long size;
+
+	private final int memorySegmentSize;
+
+	FileChannelBoundedData(
+			Path filePath,
+			FileChannel fileChannel,
+			int memorySegmentSize) {
+
+		this.filePath = checkNotNull(filePath);
+		this.fileChannel = checkNotNull(fileChannel);
+		this.memorySegmentSize = memorySegmentSize;
+		this.headerAndBufferArray = BufferReaderWriterUtil.allocatedWriteBufferArray();
+	}
+
+	@Override
+	public void writeBuffer(Buffer buffer) throws IOException {
+		size += BufferReaderWriterUtil.writeToByteChannel(fileChannel, buffer, headerAndBufferArray);
+	}
+
+	@Override
+	public void finishWrite() throws IOException {
+		fileChannel.close();
+	}
+
+	@Override
+	public Reader createReader() throws IOException {
+		checkState(!fileChannel.isOpen());
+
+		final FileChannel fc = FileChannel.open(filePath, StandardOpenOption.READ);
+		return new FileBufferReader(fc, memorySegmentSize);
+	}
+
+	@Override
+	public long getSize() {
+		return size;
+	}
+
+	@Override
+	public void close() throws IOException {
+		IOUtils.closeQuietly(fileChannel);
+		Files.delete(filePath);
+	}
+
+	// ------------------------------------------------------------------------
+
+	public static FileChannelBoundedData create(Path filePath, int memorySegmentSize) throws IOException {
+		final FileChannel fileChannel = FileChannel.open(
+				filePath, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE);
+
+		return new FileChannelBoundedData(
+				filePath,
+				fileChannel,
+				memorySegmentSize);
+	}
+
+	// ------------------------------------------------------------------------
+
+	static final class FileBufferReader implements BoundedData.Reader, BufferRecycler {
+
+		private static final int NUM_BUFFERS = 2;
+
+		private final FileChannel fileChannel;
+
+		private final ByteBuffer headerBuffer;
+
+		private final ArrayDeque<MemorySegment> buffers;
+
+		FileBufferReader(FileChannel fileChannel, int bufferSize) {
+			this.fileChannel = checkNotNull(fileChannel);
+			this.headerBuffer = BufferReaderWriterUtil.allocatedHeaderBuffer();
+			this.buffers = new ArrayDeque<>(NUM_BUFFERS);
+
+			for (int i = 0; i < NUM_BUFFERS; i++) {
+				buffers.addLast(MemorySegmentFactory.allocateUnpooledOffHeapMemory(bufferSize, null));
+			}
+		}
+
+		@Nullable
+		@Override
+		public Buffer nextBuffer() throws IOException {
+			final MemorySegment memory = buffers.pollFirst();
+
+			if (memory != null) {
+				final Buffer next = BufferReaderWriterUtil.readFromByteChannel(fileChannel, headerBuffer, memory, this);
+				if (next != null) {
+					return next;
+				}
+				else {
+					recycle(memory);
+					return null;
+				}
+			}
+
+			throw new IOException("Bug in BoundedBlockingSubpartition with FILE data: " +
```

Comment:
Though, theoretically, two read buffers is enough. But there could be still problems, because the writeAndFlush method of netty channel dose not work synchronously and the reader can be scheduled not only when the buffer of the same reader is sent out but also when the downstream adding credit or when buffers of other readers multiplex the same channel are sent out. So it is unsafe to make the assumption that when the reader is scheduled there must be a buffer for reading ahead.

Reply:
The next read should happen when the async write call is done (via channel future in netty).
The Flink 1.8 implementation has a similar assumption.
Also, without this assumption we need a theoretically unbounded buffer pool, which is not nice.

It should work in theory, but I am not 100% confident in practice. 
One more reason that I want to not expose this to users just now.

Reply:
There should be problems in practice. However, now that the FILE mode is not offered to users, we can solve the problem latter.